### PR TITLE
Completions for snap

### PR DESCRIPTION
**Pull request by radek-sprta**
## Description

Basic completions for snap. Still need to work out quirks with couple of the commands. Putting this here for feedback.

## TODOs:

- [x] Changes to fish usage are reflected in user documentation/manpages.
- [x] Tests have been added for regressions fixed
- [x] User-visible changes noted in CHANGELOG.md

### CODE REVIEW

File: share/completions/snap.fish [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+	return 1
+end
+
+function __fish_snap_subcommand
+    set subcommand $argv[1]; set -e argv[1]
+    complete -f -c snap -n '__fish_snap_no_subcommand' -a $subcommand $argv
+end
+
+function __fish_snap_disabled_snaps --description 'List disabled snaps'
+    snap list | grep disabled | cut -d ' ' -f 1
```
Comment (zanchey) [inline share/completions/snap.fish]: I'd suggest using the `string` builtin here, and throughout the completion where you use grep/cut etc.

Comment (radek-sprta) [inline share/completions/snap.fish]: Ok, switched to `string` everywhere.

**Comment by radek-sprta** [comment]
I ran into an issue with the `snap connect/disconnect` subcommands. The argument order is `snap connect snap:interface`.
I got it working to the point where it gives completions for snap and interface and results in `snap connect snap: interface`. Any idea how to get rid of the space after colon, that gets inserted after choosing a completion from the dropdown menu?

**Comment by faho** [comment]
>The argument order is snap connect snap:interface.

So you mean the `snap:interface` is one token? Fish's completion works on one token at a time. What you want here is something like

```fish
complete -f -c snap -n '__fish_snap_using_subcommand connect' -a '(__fish_snap_connectable_snaps):(__fish_snap_interfaces)'
```

The two command substitutions will be combined in a cartesian product, so this should result in the correct thing.

If this however results in too many candidates, you can use `commandline -ct` and `string match` to check if there's a ":" in the current token. However, you then need to print the complete resulting token (e.g. `printf '%s\n' (string replace -r '.*:' '' -- $token):(__fish_snap_interfaces)`).

**Comment by radek-sprta** [comment]
Thanks for the swift answer!
The Cartesian product would be too big I think (I was surprised how many interfaces were added since I last used snap), but I will try the second approach tonight.

**Comment by radek-sprta** [comment]
Should be feature complete now.

**Comment by krader1961** [comment]
Squash merged as commit 74cac0f86f0. Thx.

**Pull request closed without merging**